src/security/Handshake.cc

```diff
@@ -274,39 +274,47 @@ Security::HandshakeParser::parseModernRecord()
     Must(record.fragment.length() || record.type == ContentType::ctApplicationData);
 
     if (currentContentType != record.type) {
+        tkMessages.expectMore(false);
         parseMessages();
         Must(tkMessages.atEnd()); // no currentContentType leftovers
-        fragments = record.fragment;
+        tkMessages.reset(record.fragment, true);
         currentContentType = record.type;
     } else {
-        fragments.append(record.fragment);
+        tkMessages.reset(tkMessages.leftovers().append(record.fragment), true);
     }
 
-    if (tkRecords.atEnd() && !done)
+    if (!done)
         parseMessages();
 }
 
 /// parses one or more "higher-level protocol" frames of currentContentType
 void
 Security::HandshakeParser::parseMessages()
 {
-    tkMessages.reset(fragments, false);
-    for (; !tkMessages.atEnd(); tkMessages.commit()) {
-        switch (currentContentType) {
-        case ContentType::ctChangeCipherSpec:
-            parseChangeCipherCpecMessage();
-            continue;
-        case ContentType::ctAlert:
-            parseAlertMessage();
-            continue;
-        case ContentType::ctHandshake:
-            parseHandshakeMessage();
-            continue;
-        case ContentType::ctApplicationData:
-            parseApplicationDataMessage();
-            continue;
+    try {
+        for (; !tkMessages.atEnd(); tkMessages.expectMore(false), tkMessages.commit()) {
```

Comment:
Why do we force tkMessages.expectMore() to false after each loop iteration here?

Reply:
That is probably a leftover from the old code (setting false unconditionally). I removed it.